### PR TITLE
sniffer interfaces

### PR DESCRIPTION
**Pull request by rcohencyberarmor**
directories:
cmd : will container the main code
internal: will contain all relevant internal feature that the sniffer will use such as: config, version, etc
pkg: will contain all created packages for the sniffer feature

### CODE REVIEW

File: internal/config/config_interface.go
Comment (dwertent) [inline internal/config/config_interface.go:8]: [Question]: Why is this in the `configClient` interface? You should add a ebpfEngine interface or something like that

File: internal/config/config_interface.go
Comment (dwertent) [inline internal/config/config_interface.go:15]: [Question]: Same as before, should this function be in this interface?

File: internal/validator/validator_interface.go
Comment (dwertent) [inline internal/validator/validator_interface.go:4]: [Question]: Why is the interface private but the function public?

File: internal/version/version_interface.go
Comment (dwertent) [inline internal/version/version_interface.go:3]: [Question]: Why is the interface private but the function public?

File: internal/version/version_interface.go
Comment (dwertent) [inline internal/version/version_interface.go:4]: [Question]: How are you setting a version without passing anything?

File: cmd/main.go
Comment (dwertent) [inline cmd/main.go:1]: [Question]: I'm not sure you need a `cmd` package since the sniffer binary will never run as a CLI tool.
@matthyx What do you think about this?

File: pkg/accumulator/accumulator_interface.go
Comment (dwertent) [inline pkg/accumulator/accumulator_interface.go:4]: [Question]: Do you mean `StartAccumulatorCache`? What does this function do? please add some comments

File: pkg/accumulator/accumulator_interface.go
Comment (dwertent) [inline pkg/accumulator/accumulator_interface.go:3]: [TODO]: I'm not sure this interface name makes sense to me.

File: pkg/conthandler/container_aggregator_interface.go
Comment (dwertent) [inline pkg/conthandler/container_aggregator_interface.go:6]:
```suggestion
	ListContainerRealTimeFiles() []string
```
How can you have the `Get` and `List` in a single function name? Is it `Get` or `List`?

File: pkg/conthandler/container_main_handler.go
Comment (dwertent) [inline pkg/conthandler/container_main_handler.go:4]: [TODO]: I dont see a reason of wrapping the main function in an interface. What other options are there to run here?

File: pkg/conthandler/container_watcher_interface.go
Comment (dwertent) [inline pkg/conthandler/container_watcher_interface.go:3]: [Question]: Why is the interface private but the function public?

File: pkg/conthandler/container_watcher_interface.go
Comment (dwertent) [inline pkg/conthandler/container_watcher_interface.go:4]:
```suggestion
	WatchNewContainers()
```
[TODO]: There is no need of stating `start` and `watch` in the same function name

File: pkg/ebpfeng/ebpf_engine_interface.go
Comment (dwertent) [inline pkg/ebpfeng/ebpf_engine_interface.go:3]: [Question]: I'm not sure it is right to have the `Client` as part of an interface name

File: pkg/ebpfeng/ebpf_engine_interface.go
Comment (dwertent) [inline pkg/ebpfeng/ebpf_engine_interface.go:4]: [TODO]: I don't see any need in repeating the fact that this is `EbpfEngine`. How about `GetData` instead?

File: pkg/ebpfev/event_interface.go
Comment (dwertent) [inline pkg/ebpfev/event_interface.go:6]: [Required]: Why are the fields public?

File: pkg/ebpfev/event_interface.go
Comment (dwertent) [inline pkg/ebpfev/event_interface.go:5]: [Question]: What interface is covering this struct? Where are the getters and setters?

File: pkg/ebpfev/event_interface.go
Comment (dwertent) [inline pkg/ebpfev/event_interface.go:1]: [Required]: This package should be nested under `v1`

File: pkg/sbom/go.mod
Comment (dwertent) [inline pkg/sbom/go.mod:1]: [Required]: What?! Why?!

**Review by dwertent — CHANGES_REQUESTED** [review]
Thank you, we can start working on this project.

1. Please attach to the PR the blog you used for the package structure methodology 
2. Any data structure packages should be nested under a version. In this case - `v1`. This way, when you change the data structure you do not need to edit or delete the deprecated packages. (when moving files, please use `git mv <src> <dest>`.
3. In general: You should pay more attention to naming conventions especially for public/private.

**IMPORTANT:** Do not proceed with developing here other functions before you fix the review. We will merge this PR once you fix everything and only after you can continue developing.

**Comment by dwertent** [comment]
Please fix the [DCO](https://github.com/kubescape/sniffer/pull/2/checks?check_run_id=11119589574) as well

File: .vscode/launch.json
Comment (rcohencyberarmor) [reply .vscode/launch.json:1]: if someone want to debug it, it should help him, so I think we should stay this file here

File: install_dependencies.sh
Comment (rcohencyberarmor) [reply install_dependencies.sh:1]: why to use cmake?? cmake purpose is to create makefile

File: pkg/ebpfev/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev/event_interface.go:6]: cause I use this fields out of this package and I don't want to create getters and setters for any field

File: pkg/ebpfev/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev/event_interface.go:5]: meanwhile I don't think I need to use it has an interface, this is a place holder If I'll need in the future, too many getters and setters! I prefer to use them as public

File: README.md
Comment (rcohencyberarmor) [reply README.md:26]: will do, after implematation

File: pkg/ebpfev/event_interface.go
Comment (dwertent) [reply pkg/ebpfev/event_interface.go]: You should write getters and setters for a struct to provide encapsulation, which makes it possible to hide the internal state of an object and control access to its properties.

Getters are used to retrieve the value of a struct property, while setters are used to change the value of a struct property. These methods allow you to enforce rules and constraints on how a struct's properties can be modified and accessed, making it easier to maintain the integrity of your data and prevent bugs.

Using an interface to define the getter and setter methods can be useful in some cases, as it allows you to define a common contract that multiple structs can implement, making it easier to write generic code that works with multiple types.

File: install_dependencies.sh
Comment (dwertent) [inline install_dependencies.sh:7]: [Required]: Move this repo to Kubescape org. It is required a PR and PR approval before we use it in production

File: install_dependencies.sh
Comment (dwertent) [inline install_dependencies.sh:15]: [Question]: Is it a must to run this as an external binary? Why not import and run in a separate thread?

**Review by dwertent — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (dwertent) [inline .gitignore:3]: [Required]: Great. Now pls run `git rm .vscode`

File: .vscode/launch.json
Comment (dwertent) [reply .vscode/launch.json:1]: This is why you add documentation. In the documentation you can have a section for "Debugging on vscode"

File: pkg/ebpfev_v1/event_interface.go
Comment (dwertent) [inline pkg/ebpfev_v1/event_interface.go:1]: [Required]: `ebpfev_v1` -> `ebpfev/v1` and the package should be named  `ebpfev` (notice, `v1` is a directory nested in `ebpfev`)

Comment (dwertent) [inline pkg/ebpfev_v1/event_interface.go:1]: [Question]: What does `ebpfev` stand for?

File: pkg/sbom/sbom_interface.go
Comment (dwertent) [inline pkg/sbom/sbom_interface.go:6]:
```suggestion
	PostFilteredSBOM(key inteface{}, data []byte)
```

File: pkg/sbom/sbom_interface.go
Comment (dwertent) [inline pkg/sbom/sbom_interface.go:5]:
```suggestion
	FilterSBOM()
```

File: pkg/sbom/sbom_interface.go
Comment (dwertent) [inline pkg/sbom/sbom_interface.go:4]:
```suggestion
	GetSBOM(imageID)
```

**Review by dwertent — CHANGES_REQUESTED** [review]

File: install_dependencies.sh
Comment (rcohencyberarmor) [reply install_dependencies.sh:15]: this isn't a GO package

File: pkg/ebpfev_v1/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev_v1/event_interface.go:1]: ebpf event

File: pkg/ebpfev/v1/event_interface.go
Comment (dwertent) [inline pkg/ebpfev/v1/event_interface.go:1]: [Clarification]: I see you "put" the interface in a "version package, this is not needed because interfaces do not typically need to be in a version package. Interfaces in Go are contracts that describe the behavior of a type, rather than its structure. Because they are only contracts, changing an interface does not generally break existing code, as long as the types that implement the interface continue to satisfy the contract. If the change to an interface is backwards compatible, it can typically be made without creating a new version of the interface.

However, in some cases, it might make sense to version an interface. For example, if you are making a backwards incompatible change to the interface, creating a new version of the interface can help avoid breaking existing code. But this is not a common pattern and it should be used sparingly.

As for datastructure, having datastructure packages in a "version" package is a common pattern in Go to handle backwards compatibility issues. Go is a statically-typed language, which means that once you define a type, it is difficult to change it without breaking existing code that depends on it. This can be particularly problematic for data structures that are part of a public API, as changing the structure may require extensive modification to the calling code.

To avoid these issues, you will often define multiple versions of a data structure, each with its own package. For example, instead of defining a single "list" package, a developer might define "v1/list" and "v2/list", each with its own version of the data structure. By doing this, the developer can make breaking changes to the data structure in the "v2/list" package without affecting the existing "v1/list" package.

File: pkg/sbom/sbom_interface.go
Comment (dwertent) [inline pkg/sbom/sbom_interface.go:1]: [Required]: See my comment above regarding "nesting" datastructures in "version" packages

File: go.mod
Comment (dwertent) [inline go.mod:7]: [Required]: As I mentioned before, this package must pass code review like any other code that is released as part of Kubescape.

Please make sure to move transfer this code to the Kubescape organization or to create a fork so we can review and manage it properly.

File: docs/BUILD.md
Comment (dwertent) [inline docs/BUILD.md:1]: [Required]: I'm missing a title/header indicating what is the content of this file

File: docs/BUILD.md
Comment (dwertent) [inline docs/BUILD.md:13]: [Question]: Where is the run? 
[Statement]: We will complete the README after we complete the initial commit. No need to spend more time here.

File: README.md
Comment (dwertent) [reply README.md]: Resolved since this step was removed

File: README.md
Comment (dwertent) [inline README.md:12]: [TODO]: We should replace the env prefix.

@slashben Do we have a new name for the sniffer?

File: main.go
Comment (dwertent) [inline main.go:46]: [Required]: You will print `nil` since the error is `nil`

File: cmd/main.go
Comment (dwertent) [reply cmd/main.go]: How about moving EVERYTHING from the main function?

File: main.go
Comment (dwertent) [inline main.go:27]: [Question]: Where is this function? Do not push code that does not compile.

File: pkg/accumulator/accumulator_interface.go
Comment (dwertent) [inline pkg/accumulator/accumulator_interface.go:4]: [Question] "store the data" this is contradicting the function name. Where does it store the data? Or that by "store" you mean "cache", so don't use different words, it is confusing.

Comment (dwertent) [inline pkg/accumulator/accumulator_interface.go:4]:
```suggestion
	// StartCacheAccumulator will store the data from the ebpf engine
```
Well, you are obviously referring to the function below, no need to mention it.

File: pkg/accumulator/accumulator_interface.go
Comment (dwertent) [inline pkg/accumulator/accumulator_interface.go:3]:
```suggestion
type AccumulatorClient interface {
```
[Required]: CHECK TYPOS BEFORE YOU PUSH!!!!!!

File: pkg/sbom/sbom_interface.go
Comment (dwertent) [inline pkg/sbom/sbom_interface.go:6]: [Question]: Post? Think of a different name.

**Review by dwertent — CHANGES_REQUESTED** [review]

File: pkg/ebpfeng/ebpf_engine_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfeng/ebpf_engine_interface.go:6]: done

File: pkg/ebpfev/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev/event_interface.go]: done

File: pkg/ebpfev/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev/event_interface.go]: done

File: pkg/ebpfev/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev/event_interface.go:1]: done

File: pkg/sbom/go.mod
Comment (rcohencyberarmor) [reply pkg/sbom/go.mod]: done

File: pkg/ebpfev_v1/event_interface.go
Comment (rcohencyberarmor) [reply pkg/ebpfev_v1/event_interface.go]: done

File: pkg/sbom/sbom_interface.go
Comment (rcohencyberarmor) [reply pkg/sbom/sbom_interface.go]: done

File: pkg/sbom/sbom_interface.go
Comment (rcohencyberarmor) [reply pkg/sbom/sbom_interface.go]: done

File: pkg/sbom/sbom_interface.go
Comment (rcohencyberarmor) [reply pkg/sbom/sbom_interface.go]: done

File: go.mod
Comment (rcohencyberarmor) [reply go.mod]: I have removed the package of kubescape-ebpf from go.mod in the interface phase

File: pkg/sbom/sbom_interface.go
Comment (rcohencyberarmor) [reply pkg/sbom/sbom_interface.go]: done

File: pkg/accumulator/accumulator_interface.go
Comment (rcohencyberarmor) [reply pkg/accumulator/accumulator_interface.go]: done

File: main.go
Comment (rcohencyberarmor) [reply main.go]: done

File: main.go
Comment (rcohencyberarmor) [reply main.go]: done

File: docs/BUILD.md
Comment (rcohencyberarmor) [reply docs/BUILD.md]: done

File: README.md
Comment (matthyx) [inline README.md:9]: sneefer or sniffer?

File: README.md
Comment (matthyx) [inline README.md:36]: sneefer or sniffer?

File: install_dependencies.sh
Comment (matthyx) [inline install_dependencies.sh:1]: is it better to have a Makefile?

File: internal/config/config_interface.go
Comment (matthyx) [inline internal/config/config_interface.go:5]: should we use a io.Reader instead?

File: internal/validator/validator_interface.go
Comment (matthyx) [inline internal/validator/validator_interface.go:4]: can you use a spellchecker? I am sure there's plenty of extensions in VSCode for that.

File: internal/version/version_interface.go
Comment (matthyx) [inline internal/version/version_interface.go:4]: returns a string?

File: pkg/conthandler/container_aggregator_interface.go
Comment (matthyx) [inline pkg/conthandler/container_aggregator_interface.go:6]: no need for error here? (maybe not...)

File: pkg/conthandler/container_watcher_interface.go
Comment (matthyx) [reply pkg/conthandler/container_watcher_interface.go:4]: +1

Comment (matthyx) [reply pkg/conthandler/container_watcher_interface.go:4]: also error handling maybe?

File: pkg/sbom/sbom_interface.go
Comment (matthyx) [inline pkg/sbom/sbom_interface.go:3]: SBOMClient?

File: pkg/sbom/sbom_interface.go
Comment (matthyx) [inline pkg/sbom/sbom_interface.go:4]: error handling?

**Review by matthyx — COMMENTED** [review]
overall looks 90% good 👍 
check if you need error handling in all places, add a spellchecker to your IDE

**Review by matthyx — APPROVED** [review]

**Review by matthyx — APPROVED** [review]

**Pull request merged**
